pkg/controller/persistentvolume/persistentvolume_provisioner_controller.go

```diff
@@ -0,0 +1,530 @@
+}
+
+func (c *realControllerClient) UpdatePersistentVolumeClaim(claim *api.PersistentVolumeClaim) (*api.PersistentVolumeClaim, error) {
+	return c.client.PersistentVolumeClaims(claim.Namespace).Update(claim)
+}
+
+func (c *realControllerClient) UpdatePersistentVolumeClaimStatus(claim *api.PersistentVolumeClaim) (*api.PersistentVolumeClaim, error) {
+	return c.client.PersistentVolumeClaims(claim.Namespace).UpdateStatus(claim)
+}
+
+func (c *realControllerClient) GetKubeClient() client.Interface {
+	return c.client
+}
+
+func keyExists(key string, haystack map[string]string) bool {
+	_, exists := haystack[key]
+	return exists
+}
+
+func isAnnotationMatch(key, needle string, haystack map[string]string) bool {
+	value, exists := haystack[key]
+	if !exists {
+		return false
+	}
+	return value == needle
+}
+
+func isRecyclable(policy api.PersistentVolumeReclaimPolicy) bool {
+	return policy == api.PersistentVolumeReclaimDelete || policy == api.PersistentVolumeReclaimRecycle
+}
+
+// VolumeHost implementation
+// PersistentVolumeRecycler is host to the volume plugins, but does not actually mount any volumes.
+// Because no mounting is performed, most of the VolumeHost methods are not implemented.
+func (c *PersistentVolumeController) GetPluginDir(podUID string) string {
+	return ""
+}
+
+func (c *PersistentVolumeController) GetPodVolumeDir(podUID types.UID, pluginName, volumeName string) string {
+	return ""
+}
+
+func (c *PersistentVolumeController) GetPodPluginDir(podUID types.UID, pluginName string) string {
+	return ""
+}
+
+func (c *PersistentVolumeController) GetKubeClient() client.Interface {
+	return c.client.GetKubeClient()
+}
+
+func (c *PersistentVolumeController) NewWrapperBuilder(spec *volume.Spec, pod *api.Pod, opts volume.VolumeOptions) (volume.Builder, error) {
+	return nil, fmt.Errorf("NewWrapperBuilder not supported by PVClaimBinder's VolumeHost implementation")
+}
+
+func (c *PersistentVolumeController) NewWrapperCleaner(spec *volume.Spec, podUID types.UID) (volume.Cleaner, error) {
+	return nil, fmt.Errorf("NewWrapperCleaner not supported by PVClaimBinder's VolumeHost implementation")
+}
+
+func (c *PersistentVolumeController) GetCloudProvider() cloudprovider.Interface {
+	return c.cloud
+}
+
+func (c *PersistentVolumeController) GetMounter() mount.Interface {
+	return nil
+}
+
+func (c *PersistentVolumeController) GetWriter() io.Writer {
+	return nil
+}
+
+func (c *PersistentVolumeController) GetHostName() string {
+	return ""
+}
+
+const (
+	pvRecycleRequired  = "volume.experimental.kubernetes.io/recycle-required"
+	pvRecycleCompleted = "volume.experimental.kubernetes.io/recycle-completed"
+
+	pvDeleteRequired  = "volume.experimental.kubernetes.io/delete-required"
+	pvDeleteCompleted = "volume.experimental.kubernetes.io/delete-completed"
+
+	pvProvisioningRequired  = "volume.experimental.kubernetes.io/provisioning-required"
+	pvProvisioningCompleted = "volume.experimental.kubernetes.io/provisioning-completed"
```

Comment:
this is a bit confusing.  The way I read it, `pvProvisioningCompleted` is an annotation value that gets set in the `pvProvisioningRequired` annotation.  If that is true then it seems like `pvProvisioningRequired` is the state holder and `pvProvisioningCompleted` is a state value (another of which is `true` when it is actually required but not provisioned).

I assume that the same goes for the other constants here.